File: games/replacements_input_parse.py
# Дублирует логику static/js/replacements_input_parse.js для тестов и (при необходимости) сервера.
# При изменении правил разбора обновляйте оба файла.
# Тесты: games/tests/test_replacements_input_parse.py

import re


def norm_paste_text(s):
    return (s or '').replace('\u00a0', ' ').replace('\r\n', '\n')


def normalize_quotes_line(s):
    t = norm_paste_text(s)
    for a, b in (
        ('\u201c', '"'),
        ('\u201d', '"'),
        ('\u201e', '"'),
        ('\u2033', '"'),
        ('\u00ab', '«'),
        ('\u00bb', '»'),
    ):
        t = t.replace(a, b)
    return t


def first_non_empty_line_in_raw(raw):
    for ln in norm_paste_text(str(raw)).split('\n'):
        if ln.strip():
            return ln
    return ''


def _is_answer_token_char(ch):
    return ch.isalpha() or ch.isdigit()


def letter_token_spans(s):
    """Фрагменты из букв/цифр (как в вставке, lower для сопоставления). Пунктуация игнорируется."""
    tokens = []
    text = norm_paste_text(str(s))
    i = 0
    n = len(text)
    while i < n:
        if _is_answer_token_char(text[i]):
            j = i + 1
            while j < n and _is_answer_token_char(text[j]):
                j += 1
            orig = text[i:j]
            tokens.append((orig, orig.lower()))
            i = j
        else:
            i += 1
    return tokens


def template_token_pattern(literals, n_slots):
    """Шаблон строки: lower-слова из статического текста и None на месте каждого слота."""
    n = int(n_slots) if n_slots else 0
    if not literals or len(literals) != n + 1:
        return None
    pattern = []
    for i in range(n + 1):
        for _orig, low in letter_token_spans(literals[i]):
            pattern.append(low)
        if i < n:
            pattern.append(None)
    return pattern


def align_answers_to_template(pattern, user_spans):
    """
    Сопоставить вставку с шаблоном: литералы по lower, в слоты — следующее слово.
    Лишние слова во вставке перед литералом пропускаются.
    Литералы после последнего слота не обязательны (в шаблоне остаётся текст
    вроде BUSINESSman, в расшифровке — другое слово той же позиции).
    """
    if not pattern or not user_spans:
        return None
    n_slots = sum(1 for p in pattern if p is None)
    if len(user_spans) == n_slots:
        return [orig for orig, _low in user_spans]
    out = []
    j = 0
    n_user = len(user_spans)
    for p in pattern:
        if p is None:
            if j >= n_user:
                return None
            out.append(user_spans[j][0])
            j += 1
        else:
            if len(out) >= n_slots:
                continue
            while j < n_user and user_spans[j][1] != p:
                j += 1
            if j >= n_user or user_spans[j][1] != p:
                return None
            j += 1
    if len(out) != n_slots:
        return None
    return out


def parse_repl_token_line(raw, literals, n_slots):
    """Разбор строки: только буквы, lower, пробелы/переводы строк."""
    n = int(n_slots) if n_slots else 0
    if n <= 0:
        return []
    first = first_non_empty_line_in_raw(raw)
    if not first.strip():
        return None
    user = letter_token_spans(first)
    if not user:
        return None
    pattern = template_token_pattern(literals, n)
    if pattern is None:
        return None
    return align_answers_to_template(pattern, user)


def parse_repl_tab_line(raw, n_slots):
    n = int(n_slots) if n_slots else 0
    raw = (raw or '').strip()
    if n <= 0:
        return []
    if not raw:
        return None
    if n == 1:
        return [raw]
    tab = raw.split('\t')
    if len(tab) == n:
        return [x.strip() for x in tab]
    semi = re.split(r'\s*;\s*', raw)
    if len(semi) == n:
        return [x.strip() for x in semi]
    return None


def expand_quoted_chunks_by_whitespace(chunks, n_slots):
    """Несколько кавычечных групп → плоский список по пробелам (4 группы → 11 слотов)."""
    n = int(n_slots) if n_slots else 0
    if not chunks or n <= 0:
        return None
    flat = []
    for seg in chunks:
        seg = (seg or '').strip()
        if not seg:
            continue
        for t in seg.split():
            t = t.strip()
            if not t:
                continue
            t = re.sub(r'\?+$', '', t)
            t = re.sub(r'!+$', '', t)
            flat.append(t)
    if len(flat) == n:
        return flat
    if len(flat) > n:
        return flat[:n]
    return None


def parse_repl_mixed_quoted_line(line, n_slots):
    n = int(n_slots) if n_slots else 0
    line = normalize_quotes_line(str(line))
    if not line.strip():
        return None
    out = []
    i = 0
    ln = len(line)
    while i < ln and len(out) < n + 24:
        while i < ln and line[i].isspace():
            i += 1
        if i >= ln:
            break
        c = line[i]
        if c == '"':
            i += 1
            buf = []
            while i < ln:
                if line[i] == '\\' and i + 1 < ln:
                    buf.append(line[i + 1])
                    i += 2
                    continue
                if line[i] == '"':
                    break
                buf.append(line[i])
                i += 1
            if i < ln and line[i] == '"':
                i += 1
            out.append(''.join(buf).strip())
        elif c == '«':
            i += 1
            buf = []
            while i < ln and line[i] != '»':
                buf.append(line[i])
                i += 1
            if i < ln and line[i] == '»':
                i += 1
            out.append(''.join(buf).strip())
        else:
            i += 1
    if len(out) >= n:
        return out[:n]
    expanded = expand_quoted_chunks_by_whitespace(out, n)
    if expanded is not None:
        return expanded
    return None


def parse_repl_quoted_double_line(raw, n_slots):
    n = int(n_slots) if n_slots else 0
    line = normalize_quotes_line(str(raw)).split('\n')[0]
    if not line.strip():
        return None
    if n == 1:
        t = line.strip()
        if len(t) >= 2 and t[0] == '"' and t[-1] == '"':
            return [t[1:-1].strip()]
        return None
    out = []
    for m in re.finditer(r'"((?:[^"\\]|\\.)*)"', line):
        inner = re.sub(r'\\(.)', r'\1', m.group(1)).strip()
        out.append(inner)
    if len(out) == n:
        return out
    if len(out) > n:
        return out[:n]
    exp = expand_quoted_chunks_by_whitespace(out, n)
    if exp is not None:
        return exp
    return None


def parse_repl_guillemet_line(raw, n_slots):
    n = int(n_slots) if n_slots else 0
    line = normalize_quotes_line(str(raw)).split('\n')[0]
    if not line.strip():
        return None
    if n == 1:
        m = re.match(r'^\s*«([^»]*)»\s*$', line)
        if m:
            return [m.group(1).strip()]
        return None
    out = re.findall(r'«([^»]*)»', line)
    out = [x.strip() for x in out]
    if len(out) == n:
        return out
    if len(out) > n:
        return out[:n]
    exp2 = expand_quoted_chunks_by_whitespace(out, n)
    if exp2 is not None:
        return exp2
    return None


_OPTIONAL_SUFFIX_RE = re.compile(r'^[\s.,;:!?)»"\']+$')
_HYPHEN_LITERALS = frozenset({'-', '–', '—'})
_LONG_OPTIONAL_SUFFIX_LEN = 20


def _norm_literal_match(s):
    """Дефисы в шаблоне и вставке: -, –, — считаем одним символом."""
    return (s or '').replace('\u2013', '-').replace('\u2014', '-')


def _find_literal(rem, lit):
    if not lit:
        return 0
    pos = rem.find(lit)
    if pos >= 0:
        return pos
    return _norm_literal_match(rem).find(_norm_literal_match(lit))


def _skip_literal(rem, lit):
    if not lit:
        return rem
    if rem.startswith(lit):
        return rem[len(lit) :]
    nr = _norm_literal_match(rem)
    nl = _norm_literal_match(lit)
    if nl and nr.startswith(nl):
        return rem[len(nl) :]
    return rem


def _consume_literal_prefix(rem, lit):
    if not lit or not re.sub(r'\s', '', lit):
        return rem
    if rem.startswith(lit):
        return rem[len(lit) :]
    r0 = rem.lstrip()
    l0 = lit.lstrip()
    if l0 and r0.startswith(l0):
        return rem[len(rem) - len(r0) + len(l0) :]
    nr = _norm_literal_match(rem.lstrip())
    nl = _norm_literal_match(lit.lstrip())
    if nl and nr.startswith(nl):
        return rem[len(rem.lstrip()) - len(nr) + len(nl) :]
    return None


def _slice_before_suffix(rem, after):
    """Значение перед хвостовым литералом; пунктуацию в suffix можно не копировать."""
    if not after:
        return rem.strip()
    if rem.endswith(after):
        return rem[:-len(after)].strip()
    na = _norm_literal_match(after)
    nr = _norm_literal_match(rem)
    if na and nr.endswith(na):
        return rem[: len(rem) - len(after)].strip() if rem.endswith(after) else rem[
            : len(nr) - len(na)
        ].strip()
    if _OPTIONAL_SUFFIX_RE.match(after):
        r = rem.rstrip()
        suf = after.strip()
        while suf:
            if r.endswith(suf):
                return r[: -len(suf)].strip()
            suf = suf[:-1]
        return r.strip()
    tr = rem.rstrip()
    ta = after.rstrip()
    if tr.endswith(ta):
        return tr[: -len(ta)].strip()
    return None


def _next_significant_literal(literals, start_idx):
    for j in range(start_idx, len(literals)):
        lit = literals[j]
        if lit and lit not in _HYPHEN_LITERALS:
            return lit
    return ''


def _try_hyphen_pair_without_char(rem, next_lit):
    pos = _find_literal(rem, next_lit)
    if pos < 0:
        return None
    words = rem[:pos].split()
    if len(words) != 2:
        return None
    return words[0], words[1], rem[pos:]


def parse_repl_compact_fallback_line(first_line, n_slots):
    """Строка только из ответов: ровно N буквенных токенов."""
    n = int(n_slots) if n_slots else 0
    if n <= 0:
        return []
    user = letter_token_spans(first_line or '')
    if len(user) == n:
        return [orig for orig, _low in user]
    return None


def parse_full_line_by_literals(full_raw, literals, n_slots):
    """Разбор целой строки ответа по литералам из DOM (режим «по словам» / «весь текст»)."""
    n = int(n_slots) if n_slots else 0
    if n <= 0:
        return []
    if not literals or len(literals) != n + 1:
        return None
    first = str(full_raw).split('\n')[0].rstrip()
    rem = _consume_literal_prefix(first, literals[0])
    if rem is None:
        return None
    out = []
    i = 0
    while i < n:
        if i == n - 1:
            after = literals[n] or ''
            val = _slice_before_suffix(rem, after)
            if val is None and len(after.strip()) >= _LONG_OPTIONAL_SUFFIX_LEN:
                val = rem.strip()
            if val is None:
                return None
            out.append(val)
            break
        nl = literals[i + 1]
        if nl in _HYPHEN_LITERALS:
            pos = _find_literal(rem, nl)
            if pos >= 0:
                out.append(rem[:pos].strip())
                rem = _skip_literal(rem[pos:], nl)
                i += 1
                continue
            next_lit = _next_significant_literal(literals, i + 2)
            if not next_lit:
                return None
            pair = _try_hyphen_pair_without_char(rem, next_lit)
            if pair is None:
                return None
            out.extend([pair[0], pair[1]])
            rem = pair[2]
            i += 2
            continue
        if nl == '':
            return None
        pos = _find_literal(rem, nl)
        if pos < 0:
            return None
        out.append(rem[:pos].strip())
        rem = _skip_literal(rem[pos:], nl)
        i += 1
    if len(out) != n:
        return None
    return out


def literals_from_right_tokens(line_tokens):
    """Список литералов между слотами (как extractReplWordsLiterals в JS)."""
    L = []
    buf = ''
    for tok in line_tokens:
        if tok.get('type') == 'text':
            buf += tok.get('text') or ''
        elif tok.get('type') == 'slot':
            L.append(buf)
            buf = ''
    L.append(buf)
    return L


def parse_repl_line_answers_smart(raw, n_slots, literals=None):
    """Tab/; → сопоставление буквенных токенов (без пунктуации, lower)."""
    raw_norm = norm_paste_text(raw)
    first_line = first_non_empty_line_in_raw(raw_norm).rstrip()
    r = parse_repl_line_answers_smart_no_dom(raw, n_slots)
    if r is not None:
        return r
    if literals is not None:
        r = parse_repl_token_line(raw, literals, n_slots)
        if r is not None:
            return r
    return parse_repl_compact_fallback_line(first_line, n_slots)


def parse_repl_line_answers_smart_no_dom(raw, n_slots):
    """Без разбора по литералам из DOM (только таб/кавычки/«»)."""
    if n_slots <= 0:
        return []
    raw = norm_paste_text(raw)
    if not re.sub(r'\s', '', raw):
        return None
    first_line = first_non_empty_line_in_raw(raw).rstrip()
    tabbed = parse_repl_tab_line(first_line, n_slots)
    if tabbed is not None:
        return tabbed
    mixed = parse_repl_mixed_quoted_line(first_line, n_slots)
    if mixed is not None:
        return mixed
    qd = parse_repl_quoted_double_line(first_line, n_slots)
    if qd is not None:
        return qd
    qg = parse_repl_guillemet_line(first_line, n_slots)
    if qg is not None:
        return qg
    return None

File: static/js/replacements_input_parse.js
/**
 * Разбор ввода для заданий «Замены» (строка / вставка из документа).
 * Зеркало логики: games/replacements_input_parse.py (тесты Django).
 * Локальная проверка: node static/js/replacements_input_parse.test.js
 */
(function (global) {
  'use strict';

  function normPasteText(s) {
    return (s || '').replace(/\u00a0/g, ' ').replace(/\r\n/g, '\n');
  }

  function normalizeQuotesLine(s) {
    return normPasteText(s)
      .replace(/\u201C|\u201D|\u201E|\u2033/g, '"')
      .replace(/\u00AB/g, '«')
      .replace(/\u00BB/g, '»');
  }

  function firstNonEmptyLineInRaw(raw) {
    var lines = normPasteText(String(raw)).split(/\r?\n/);
    for (var i = 0; i < lines.length; i++) {
      if (lines[i].replace(/\s/g, '')) return lines[i];
    }
    return '';
  }

  /** Буквы и цифры (латиница + кириллица); пунктуация не входит в токен. */
  var LETTER_RUN_RE = /[A-Za-z0-9\u0400-\u04FF\u0500-\u052F]+/g;

  function letterTokenSpans(s) {
    var text = normPasteText(String(s));
    var out = [];
    var m;
    LETTER_RUN_RE.lastIndex = 0;
    while ((m = LETTER_RUN_RE.exec(text)) !== null) {
      out.push({ orig: m[0], low: m[0].toLowerCase() });
    }
    return out;
  }

  function templateTokenPattern(L, nSlots) {
    if (!L || L.length !== nSlots + 1) return null;
    var pattern = [];
    var i;
    for (i = 0; i <= nSlots; i++) {
      var spans = letterTokenSpans(L[i]);
      for (var t = 0; t < spans.length; t++) pattern.push(spans[t].low);
      if (i < nSlots) pattern.push(null);
    }
    return pattern;
  }

  function alignAnswersToTemplate(pattern, userSpans) {
    if (!pattern || !userSpans || !userSpans.length) return null;
    var nSlots = 0;
    var si;
    for (si = 0; si < pattern.length; si++) {
      if (pattern[si] === null) nSlots++;
    }
    if (userSpans.length === nSlots) {
      var only = [];
      for (si = 0; si < userSpans.length; si++) only.push(userSpans[si].orig);
      return only;
    }
    var out = [];
    var j = 0;
    for (si = 0; si < pattern.length; si++) {
      if (pattern[si] === null) {
        if (j >= userSpans.length) return null;
        out.push(userSpans[j].orig);
        j++;
      } else {
        while (j < userSpans.length && userSpans[j].low !== pattern[si]) j++;
        if (j >= userSpans.length || userSpans[j].low !== pattern[si]) return null;
        j++;
      }
    }
    return out.length === nSlots ? out : null;
  }

  function parseReplTokenLine(raw, L, nSlots) {
    if (nSlots <= 0) return [];
    var first = firstNonEmptyLineInRaw(raw);
    if (!String(first).replace(/\s/g, '')) return null;
    var user = letterTokenSpans(first);
    if (!user.length) return null;
    var pattern = templateTokenPattern(L, nSlots);
    if (!pattern) return null;
    return alignAnswersToTemplate(pattern, user);
  }

  function parseReplTabLine(raw, nSlots) {
    var n = parseInt(nSlots, 10) || 0;
    raw = (raw || '').trim();
    if (n <= 0) return [];
    if (!raw) return null;
    if (n === 1) return [raw];
    var tab = raw.split('\t');
    if (tab.length === n) return tab.map(function (s) { return s.trim(); });
    var semi = raw.split(/\s*;\s*/);
    if (semi.length === n) return semi.map(function (s) { return s.trim(); });
    return null;
  }

  /**
   * Несколько кавычек с группами слов ("А Б", "В") → плоский список слотов по пробелам.
   * Нужно для строки «как в ответе»: 4 кавычки при 11 слотах в условии (капс внутри групп).
   */
  function expandQuotedChunksByWhitespace(chunks, nSlots) {
    var n = parseInt(nSlots, 10) || 0;
    if (!chunks || !chunks.length || n <= 0) return null;
    var flat = [];
    for (var c = 0; c < chunks.length; c++) {
      var seg = (chunks[c] || '').trim();
      if (!seg) continue;
      var parts = seg.split(/\s+/);
      for (var p = 0; p < parts.length; p++) {
        var t = parts[p].trim();
        if (!t) continue;
        t = t.replace(/\?+$/g, '').replace(/!+$/g, '');
        flat.push(t);
      }
    }
    if (flat.length === n) return flat;
    if (flat.length > n) return flat.slice(0, n);
    return null;
  }

  /**
   * Подряд идущие фрагменты в "..." или «...» (как в документах с « и » между группами).
   */
  function parseReplMixedQuotedLine(line, nSlots) {
    var n = parseInt(nSlots, 10) || 0;
    if (n <= 0) return [];
    line = normalizeQuotesLine(String(line));
    if (!line.replace(/\s/g, '')) return null;
    var out = [];
    var i = 0;
    var len = line.length;
    while (i < len && out.length < n + 24) {
      while (i < len && /\s/.test(line.charAt(i))) i++;
      if (i >= len) break;
      var c = line.charAt(i);
      if (c === '"') {
        i++;
        var buf = '';
        while (i < len) {
          if (line.charAt(i) === '\\' && i + 1 < len) {
            buf += line.charAt(i + 1);
            i += 2;
            continue;
          }
          if (line.charAt(i) === '"') break;
          buf += line.charAt(i);
          i++;
        }
        if (i < len && line.charAt(i) === '"') i++;
        out.push(buf.trim());
      } else if (c === '«') {
        i++;
        var buf2 = '';
        while (i < len && line.charAt(i) !== '»') {
          buf2 += line.charAt(i);
          i++;
        }
        if (i < len && line.charAt(i) === '»') i++;
        out.push(buf2.trim());
      } else {
        i++;
      }
    }
    if (out.length >= n) return out.slice(0, n);
    var expanded = expandQuotedChunksByWhitespace(out, n);
    if (expanded) return expanded;
    return null;
  }

  function parseReplQuotedDoubleLine(raw, nSlots) {
    var n = parseInt(nSlots, 10) || 0;
    if (n <= 0) return [];
    var line = normalizeQuotesLine(String(raw)).split(/\r?\n/)[0];
    if (!line.replace(/\s/g, '')) return null;
    if (n === 1) {
      var t = line.trim();
      if (t.length >= 2 && t.charAt(0) === '"' && t.charAt(t.length - 1) === '"') {
        return [t.slice(1, -1).trim()];
      }
      return null;
    }
    var re = /"((?:[^"\\]|\\.)*)"/g;
    var out = [];
    var m;
    while ((m = re.exec(line)) !== null) {
      out.push(m[1].replace(/\\(.)/g, '$1').trim());
    }
    if (out.length === n) return out;
    if (out.length > n) return out.slice(0, n);
    var exp = expandQuotedChunksByWhitespace(out, n);
    if (exp) return exp;
    return null;
  }

  function parseReplGuillemetLine(raw, nSlots) {
    var n = parseInt(nSlots, 10) || 0;
    if (n <= 0) return [];
    var line = normalizeQuotesLine(String(raw)).split(/\r?\n/)[0];
    if (!line.replace(/\s/g, '')) return null;
    if (n === 1) {
      var mm = line.match(/^\s*«([^»]*)»\s*$/);
      if (mm) return [mm[1].trim()];
      return null;
    }
    var re = /«([^»]*)»/g;
    var out = [];
    var m;
    while ((m = re.exec(line)) !== null) {
      out.push(m[1].trim());
    }
    if (out.length === n) return out;
    if (out.length > n) return out.slice(0, n);
    var exp2 = expandQuotedChunksByWhitespace(out, n);
    if (exp2) return exp2;
    return null;
  }

  function extractReplWordsLiterals(container) {
    var L = [];
    var buf = '';
    var nSlots = 0;
    if (!container || !container.children) return { literals: [], nSlots: 0 };
    var kids = container.children;
    for (var i = 0; i < kids.length; i++) {
      var el = kids[i];
      if (el.classList && el.classList.contains('new-replacements-static')) {
        buf += normPasteText(el.textContent);
      } else if (el.classList && el.classList.contains('new-replacements-partial-answer')) {
        buf += normPasteText(el.textContent);
      } else if (el.tagName === 'INPUT' && el.name === 'answers[]') {
        L.push(buf);
        buf = '';
        nSlots++;
      }
    }
    L.push(buf);
    return { literals: L, nSlots: nSlots };
  }

  function normLiteralMatch(s) {
    return String(s || '').replace(/\u2013/g, '-').replace(/\u2014/g, '-');
  }

  function findLiteral(rem, lit) {
    if (!lit) return 0;
    var pos = rem.indexOf(lit);
    if (pos >= 0) return pos;
    return normLiteralMatch(rem).indexOf(normLiteralMatch(lit));
  }

  function skipLiteral(rem, lit) {
    if (!lit) return rem;
    if (rem.indexOf(lit) === 0) return rem.slice(lit.length);
    var nr = normLiteralMatch(rem);
    var nl = normLiteralMatch(lit);
    if (nl && nr.indexOf(nl) === 0) return rem.slice(nl.length);
    return rem;
  }

  function consumeLiteralPrefix(rem, lit) {
    if (!lit || !lit.replace(/\s/g, '')) return rem;
    if (rem.indexOf(lit) === 0) return rem.slice(lit.length);
    var r0 = rem.replace(/^\s+/, '');
    var l0 = lit.replace(/^\s+/, '');
    if (l0 && r0.indexOf(l0) === 0) {
      return rem.slice(rem.length - r0.length + l0.length);
    }
    var nr = normLiteralMatch(rem.replace(/^\s+/, ''));
    var nl = normLiteralMatch(lit.replace(/^\s+/, ''));
    if (nl && nr.indexOf(nl) === 0) {
      return rem.slice(rem.length - rem.replace(/^\s+/, '').length + nl.length);
    }
    return null;
  }

  var OPTIONAL_SUFFIX_RE = /^[\s.,;:!?)»"']+$/;
  var HYPHEN_LITERALS = { '-': true, '\u2013': true, '\u2014': true };
  var LONG_OPTIONAL_SUFFIX_LEN = 20;

  function sliceBeforeSuffix(rem, after) {
    if (!after) return rem.trim();
    if (rem.endsWith(after)) return rem.slice(0, rem.length - after.length).trim();
    var na = normLiteralMatch(after);
    var nr = normLiteralMatch(rem);
    if (na && nr.endsWith(na)) {
      return rem.slice(0, rem.length - (rem.endsWith(after) ? after.length : na.length)).trim();
    }
    if (OPTIONAL_SUFFIX_RE.test(after)) {
      var r = rem.replace(/\s+$/, '');
      var suf = after.replace(/^\s+|\s+$/g, '');
      while (suf.length) {
        if (r.endsWith(suf)) return r.slice(0, r.length - suf.length).trim();
        suf = suf.slice(0, -1);
      }
      return r.trim();
    }
    var tr = rem.replace(/\s+$/, '');
    var ta = after.replace(/\s+$/, '');
    if (tr.endsWith(ta)) return tr.slice(0, tr.length - ta.length).trim();
    return null;
  }

  function nextSignificantLiteral(L, startIdx) {
    for (var j = startIdx; j < L.length; j++) {
      if (L[j] && !HYPHEN_LITERALS[L[j]]) return L[j];
    }
    return '';
  }

  function parseReplCompactFallbackLine(firstLine, nSlots) {
    var user = letterTokenSpans(firstLine || '');
    if (user.length === nSlots) {
      var out = [];
      for (var i = 0; i < user.length; i++) out.push(user[i].orig);
      return out;
    }
    return null;
  }

  function tryHyphenPairWithoutChar(rem, nextLit) {
    var pos = findLiteral(rem, nextLit);
    if (pos < 0) return null;
    var words = rem.slice(0, pos).trim().split(/\s+/);
    if (words.length !== 2) return null;
    return { w0: words[0], w1: words[1], rest: rem.slice(pos) };
  }

  function parseFullLineByLiterals(fullRaw, L, nSlots) {
    if (!nSlots) return [];
    if (!L || L.length !== nSlots + 1) return null;
    var first = String(fullRaw).split(/\r?\n/)[0].replace(/\s+$/, '');
    var rem = consumeLiteralPrefix(first, L[0]);
    if (rem === null) return null;
    var out = [];
    var i = 0;
    while (i < nSlots) {
      if (i === nSlots - 1) {
        var afterLast = L[nSlots] || '';
        var val = sliceBeforeSuffix(rem, afterLast);
        if (val === null && afterLast.replace(/^\s+|\s+$/g, '').length >= LONG_OPTIONAL_SUFFIX_LEN) {
          val = rem.replace(/^\s+|\s+$/g, '');
        }
        if (val === null) return null;
        out.push(val);
        break;
      }
      var nextLit = L[i + 1];
      if (HYPHEN_LITERALS[nextLit]) {
        var hpos = findLiteral(rem, nextLit);
        if (hpos >= 0) {
          out.push(rem.slice(0, hpos).trim());
          rem = skipLiteral(rem.slice(hpos), nextLit);
          i += 1;
          continue;
        }
        var sigLit = nextSignificantLiteral(L, i + 2);
        if (!sigLit) return null;
        var pair = tryHyphenPairWithoutChar(rem, sigLit);
        if (!pair) return null;
        out.push(pair.w0);
        out.push(pair.w1);
        rem = pair.rest;
        i += 2;
        continue;
      }
      if (nextLit === '') return null;
      var pos = findLiteral(rem, nextLit);
      if (pos < 0) return null;
      out.push(rem.slice(0, pos).trim());
      rem = skipLiteral(rem.slice(pos), nextLit);
      i += 1;
    }
    if (out.length !== nSlots) return null;
    return out;
  }

  function splitWholeLines(text) {
    if (text == null || text === '') return [];
    return text.split(/\r?\n/);
  }

  function parseReplLineAnswersSmart(container, raw, nSlots) {
    if (nSlots <= 0) return [];
    raw = normPasteText(raw);
    if (!String(raw).replace(/\s/g, '')) return null;
    var firstLine = firstNonEmptyLineInRaw(raw).replace(/\s+$/, '');
    var tabbed = parseReplTabLine(firstLine, nSlots);
    if (tabbed) return tabbed;
    var mixed = parseReplMixedQuotedLine(firstLine, nSlots);
    if (mixed) return mixed;
    var qd = parseReplQuotedDoubleLine(firstLine, nSlots);
    if (qd) return qd;
    var qg = parseReplGuillemetLine(firstLine, nSlots);
    if (qg) return qg;
    if (container) {
      var ex = extractReplWordsLiterals(container);
      if (ex.nSlots === nSlots) {
        var tok = parseReplTokenLine(firstLine, ex.literals, nSlots);
        if (tok) return tok;
      }
    }
    return parseReplCompactFallbackLine(firstLine, nSlots);
  }

  var api = {
    splitWholeLines: splitWholeLines,
    expandQuotedChunksByWhitespace: expandQuotedChunksByWhitespace,
    normPasteText: normPasteText,
    normalizeQuotesLine: normalizeQuotesLine,
    firstNonEmptyLineInRaw: firstNonEmptyLineInRaw,
    parseReplTabLine: parseReplTabLine,
    parseReplMixedQuotedLine: parseReplMixedQuotedLine,
    parseReplQuotedDoubleLine: parseReplQuotedDoubleLine,
    parseReplGuillemetLine: parseReplGuillemetLine,
    extractReplWordsLiterals: extractReplWordsLiterals,
    letterTokenSpans: letterTokenSpans,
    templateTokenPattern: templateTokenPattern,
    alignAnswersToTemplate: alignAnswersToTemplate,
    parseReplTokenLine: parseReplTokenLine,
    parseFullLineByLiterals: parseFullLineByLiterals,
    sliceBeforeSuffix: sliceBeforeSuffix,
    parseReplLineAnswersSmart: parseReplLineAnswersSmart,
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = api;
  }
  global.ReplacementsInputParse = api;
})(typeof window !== 'undefined' ? window : global);
